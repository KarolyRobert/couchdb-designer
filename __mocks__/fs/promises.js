
'use strict';

const promises = jest.createMockFromModule('fs/promises');

promises.readFile = jest.fn();
promises.writeFile = jest.fn();
promises.readdir = jest.fn();
promises.access = jest.fn();
promises.stat = jest.fn();

module.exports = promises;
