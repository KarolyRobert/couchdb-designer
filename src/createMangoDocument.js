import fs from 'fs/promises';
import path from 'path';

const mapFields = (fields,fileName,indexName) => {
    let fieldsObject = {};
    let direction = false;
    for(let field of fields) {
        if(typeof field === 'string'){
            if(direction && direction !== 'asc'){
                throw(`(unsupported_mixed_sort) Sorts currently only support a single direction for all fields. You can fix it in ${fileName} - index: ${indexName}.`);
            }else if(!direction){
                direction = 'asc';
            }
            fieldsObject[field] = direction;
        }else if(typeof field === 'object'){
            let fieldName = Object.keys(field);
            if(fieldName.length === 1 && (field[fieldName[0]] === 'asc' || field[fieldName[0]] === 'desc')){
                if(direction && direction !== field[fieldName[0]]){
                    throw(`(unsupported_mixed_sort) Sorts currently only support a single direction for all fields. You can fix it in ${fileName} - index: ${indexName}.`);
                }else if(!direction){
                    direction = field[fieldName[0]];
                }
                fieldsObject[fieldName[0]] = direction;
            }else{
                throw(`(invalid_sort_field) Invalid sort field: ${fieldName[0]}:${field[fieldName[0]]}. You can fix it in ${fileName},${indexName} index.`);
            }
        }else{
            throw(`(invalid_sort_field) Invalid sort field: ${field}. You can fix it in ${fileName},${indexName} index.`);
        }
    }
    return fieldsObject;
}


const createMangoDocument = (root,name) => {
    return new Promise((resolve, reject) => {
        let ddoc = {
            _id:`_design/${name.split('.')[0]}`,
            language:'query',
            views:{}
        }
        let file = path.join(root,name);
        fs.readFile(file).then(fileContent => {
            try{
                let json = JSON.parse(fileContent);
                let keys = Object.keys(json);
                for(let key of keys) {
                    if(key === 'partitioned'){
                        if(json.partitioned){
                            ddoc.options = { partitioned: true }
                        }else{
                            ddoc.options = { partitioned: false }
                        }
                    }else{
                        let view = { 
                            map:{
                                fields: mapFields(json[key].fields,file,key),
                                partial_filter_selector: json[key].partial_filter_selector ? json[key].partial_filter_selector : {}
                            },
                            reduce:'_count',
                            options: { 
                                def: json[key]
                            }
                        }
                        ddoc.views[key] = view;
                    }
                }
                resolve(ddoc);
            }catch(err){
                reject(err);
            }
        },reject);
    })
}

export default createMangoDocument;