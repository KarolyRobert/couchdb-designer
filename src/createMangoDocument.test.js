import createMangoDocument from "./createMangoDocument";
import fs from 'fs/promises';


jest.mock('fs/promises');

const jsonSuccess = JSON.stringify({
    "partitioned":false,
    "mongo":{
        "partial_filter_selector": {
            "year": {
                "$gt": 2010
            },
            "limit": 10,
            "skip": 0
        },
        "fields": [
            "_id",
            {"_rev":"asc"},
            "year",
            "title"
        ]
    } 
});

const jsonWrongSort1 = JSON.stringify({
    "partitioned":false,
    "mongo":{
        "partial_filter_selector": {
            "year": {
                "$gt": 2010
            },
            "limit": 10,
            "skip": 0
        },
        "fields": [
            "_id",
            {"_rev":"desc"},
            "year",
            "title"
        ]
    } 
});

const jsonWrongSort2 = JSON.stringify({
    "partitioned":false,
    "mongo":{
        "partial_filter_selector": {
            "year": {
                "$gt": 2010
            },
            "limit": 10,
            "skip": 0
        },
        "fields":[
            {"_rev":"desc"},
            "_id",
            "year",
            "title"
        ]
    } 
});


const jsonInvalidSort1 = JSON.stringify({
    "partitioned":false,
    "mongo":{
        "partial_filter_selector": {
            "year": {
                "$gt": 2010
            },
            "limit": 10,
            "skip": 0
        },
        "fields":[
            {"_rev":"desac"},
            "_id",
            "year",
            "title"
        ]
    } 
});


const jsonInvalidSort2 = JSON.stringify({
    "partitioned":false,
    "mongo":{
        "partial_filter_selector": {
            "year": {
                "$gt": 2010
            },
            "limit": 10,
            "skip": 0
        },
        "fields":[
            true,
            "_id",
            "year",
            "title"
        ]
    } 
});


const jsonInvalidSort3 = JSON.stringify({
    "partitioned":false,
    "mongo":{
        "partial_filter_selector": {
            "year": {
                "$gt": 2010
            },
            "limit": 10,
            "skip": 0
        },
        "fields":[
            {"helo":"asc","szia":"desc"},
            "_id",
            "year",
            "title"
        ]
    } 
});

describe('createMangoDocument',() => {
    beforeEach(() => {
        jest.clearAllMocks();
    });

    test('success', () => {
        fs.readFile.mockResolvedValue(jsonSuccess);
        return createMangoDocument('root','name').then(result => {
            expect(fs.readFile.mock.calls[0][0]).toBe('root/name');
            expect(result).toMatchSnapshot();
        });
    });


    test('wrong sort1', () => {
        fs.readFile.mockResolvedValue(jsonWrongSort1);
        return createMangoDocument('root','name').catch(error => {
            expect(error).toBe('(unsupported_mixed_sort) Sorts currently only support a single direction for all fields. You can fix it in root/name - index: mongo.');
        });
    });

    test('wrong sort2', () => {
        fs.readFile.mockResolvedValue(jsonWrongSort2);
        return createMangoDocument('root','name').catch(error => {
            expect(error).toBe('(unsupported_mixed_sort) Sorts currently only support a single direction for all fields. You can fix it in root/name - index: mongo.');
        });
    });

    test('invalid sort1', () => {
        fs.readFile.mockResolvedValue(jsonInvalidSort1);
        return createMangoDocument('root','name').catch(error => {
            expect(error).toBe('(invalid_sort_field) Invalid sort field: _rev:desac. You can fix it in root/name,mongo index.');
        });
    });

    test('invalid sort2', () => {
        fs.readFile.mockResolvedValue(jsonInvalidSort2);
        return createMangoDocument('root','name').catch(error => {
            expect(error).toBe('(invalid_sort_field) Invalid sort field: true. You can fix it in root/name,mongo index.');
        });
    });

    test('invalid sort3', () => {
        fs.readFile.mockResolvedValue(jsonInvalidSort3);
        return createMangoDocument('root','name').catch(error => {
            expect(error).toBe('(invalid_sort_field) Invalid sort field: helo:asc. You can fix it in root/name,mongo index.');
        });
    });

});