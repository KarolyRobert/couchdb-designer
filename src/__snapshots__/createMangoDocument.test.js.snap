// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createMangoDocument success 1`] = `
Object {
  "_id": "_design/name",
  "language": "query",
  "options": Object {
    "partitioned": false,
  },
  "views": Object {
    "mongo": Object {
      "map": Object {
        "fields": Object {
          "_id": "asc",
          "_rev": "asc",
          "title": "asc",
          "year": "asc",
        },
        "partial_filter_selector": Object {
          "limit": 10,
          "skip": 0,
          "year": Object {
            "$gt": 2010,
          },
        },
      },
      "options": Object {
        "def": Object {
          "fields": Array [
            "_id",
            Object {
              "_rev": "asc",
            },
            "year",
            "title",
          ],
          "partial_filter_selector": Object {
            "limit": 10,
            "skip": 0,
            "year": Object {
              "$gt": 2010,
            },
          },
        },
      },
      "reduce": "_count",
    },
  },
}
`;
